File: sa_scheduling/cost_functions.py
import math
import copy 
from functools import reduce

"""
    TODO do not sort but find min element 
    TODO skip to next ready job when idle. this is almost done 
    TODO Extension 1 utilization thing in edf 
"""


def get_ready(task_set, cycle, ready_list):
    for task in task_set:
        if cycle % task.period == 0:
            new_job = copy.deepcopy(task)
            new_job.deadline = new_job.deadline + cycle
            new_job.release_time = cycle
            ready_list.append(new_job)
    
    # sort on deadline
    # https://www.geeksforgeeks.org/sorting-objects-of-user-defined-class-in-python/
    ready_list = sorted(ready_list, key=lambda t: t.deadline) 
    
    return ready_list

# microticks to get to next point at which some task will be relased
def steps_to_next_release(cycle, task_set):
    return min([task.period - (cycle % task.period) for task in task_set])
        
# ts is set of TT tasks
def edf(ts): 
    periods = [t.period for t in ts]
    T = math.lcm(*periods) # least common multiple of TT task periods 
    s = [] # schedule will be hyperperiod long. 12 000 microticks == 120 000 microsecs == 120 ms
    ready_list = []
    wcrts = {} # worst case response times
    
    # overvej reset funktioner i task klassen istedet for alle de kopier...
    for t in range(0, T):
        for task in ready_list:
            if task.duration > 0 and task.deadline <= t: 
                return s, -1 # just return 1 here maybe lol? 
        
            # job done check response time gt wcrt and remove from ready list
            if task.duration == 0 and task.deadline >= t:
                response_time = t - task.release_time
                
                if task.name not in wcrts or response_time >= wcrts[task.name]:
                    wcrts[task.name] = response_time
                
                ready_list.remove(task)
                
        # release taks at time t
        ready_list = get_ready(ts, t, ready_list)

        if ready_list == []: # TODO add skip to next released 
            s.append("IDLE")
            continue
        else:
            # EDF get next job to execute 
            s.append(ready_list[0].name)
            ready_list[0].duration = ready_list[0].duration - 1
    
    if ready_list != []:
        return [], wcrts
    
    #print("in EDF wcrts is: ", wcrts)
    
    return s, wcrts

# utility function 
def unpack(task):
    return (task.priority, task.duration, task.period, task.deadline)

#Schedulability of ET tasks under a given polling task
def calculate_schedulabiltiy(polling_server):
    et_tasks = polling_server.et_subset
    Tp = polling_server.period
    Dp = polling_server.deadline
    Cp = polling_server.duration
    result_dict = {}
    is_schedulable = True
    #compute delta and alpha accordingly to [2]
    Delta = Tp + Dp - 2*Cp
    alpha = Cp/Tp
    
    #hyperperiod is lcm of all task periods in T_ET (all values must be from the chosen subset of ET tasks from the .csv)
    periods = [t.period for t in et_tasks]
     
    hyperperiod = math.lcm(*periods)
    
    for et_task in et_tasks:
        (pi, Ci, Ti, Di) = unpack(et_task)
        t = 0
        #initialize the response time of ti (task period) to a value exceeding the deadline
        response_time = Di + 1

        #remember we are dealing with constrained deadline tasks for the AdvPoll, hence, in the worst case arrival pattern, the intersection must lie within the hyperperiod if the task is schedulable

        while t <= hyperperiod:
            #the supply at time t ([1])
            supply = alpha*(t-Delta)

            #compute the maximum demand at time t according to Eq. 2
            demand = 0
            for tj in et_tasks:
                (pj, Cj, Tj, Dj) = unpack(tj)

                if pj >= pi:
                    demand  = demand + math.ceil(t/Tj)*Cj
            
            #According to lemma 1 of [1], we are searching for the earliest time, when the supply exceeds the demand
            if supply >= demand:
                response_time = t
                result_dict[et_task.name]  = (response_time, et_task.deadline) # if actually greater than deadline, set to false later
                break
            
            t = t + 1
        
        if response_time > Di: 
            is_schedulable = False # TODO maybe just return here ... and penalize with 1 
            result_dict[et_task.name] = (response_time, et_task.deadline)
            
        
    return is_schedulable, result_dict # contains wcrtbool indicating schedulability and deadline for each et



# cost is sum of worst case response times of tt tasks + sum worst case respone times for et tasks
# the wcrt of tt/et task i is multiplied by 1/deadline_i to normalize it
# the sum of wcrts tt is multiplied by 1/len(tt_task_set) and 1/len(et_task_set) for wcrts of et tasks
# this gives us a number betwen 0 and 1 for both tt and et 
# add these two together and get a number between 0 and 2
# use this as cost 
def cost_f(task_set):
    polling_servers = [ps for ps in task_set if ps.et_subset != None] # get set of polling servers from task set 
    
    l = [calculate_schedulabiltiy(ps) for ps in polling_servers] # check schedulability for each polling server
    
    wcrts_et = 0 # use worst case response time for cost metric
    is_schedulable = True # if some ps is not schedulable at some penalty to cost 
    
    # costs is (sum(wcrt_i/deadline_i) / len(et_subset)) /len(polling_servers)
    for element in l: 
        is_schedulable, wcrts = element
        if is_schedulable:
            wcrts_et += sum([wcrts[key][0] / wcrts[key][1] for key in wcrts]) / len(wcrts)
        else:
            wcrts_et += 1
        #is_schedulable = is_schedulable and reduce((lambda a, b : a and b), [entry[key][0] for key in entry]) # https://www.geeksforgeeks.org/reduce-in-python/ 
    
    # normalize such that 0 <= cost <= 1. this check is funny  
    if l != []:
        wcrts_et *= 1/len(l)

    # larger penalty maybe, maybe good idea?
    #if not is_schedulable:
    #    wcrts_et = 1

    # apply earliest deadline first 
    s, wcrts = edf(task_set)
    
    # if not tt tasks not schedulable (-1) return faulty schedule, 2 (max cost the way we normalize right now)
    # and false (not schedulable)
    if wcrts == -1:
        #return s, 2, False
        #return s, 1 + wcrts_et, False
        is_schedulable = False
        sum_wcrts_tt = 1
    else:
        # normalize worst case response time. for each tt task 0 <= wcrt <= 1 by setting it to wcrt/deadline
        sum_wcrts_tt = sum([wcrts[task.name] / task.deadline for task in task_set]) / len(task_set)
     
    print("cost et: ", wcrts_et, " cost tt: ", sum_wcrts_tt) 
    sum_wcrts = (sum_wcrts_tt + wcrts_et)

    #alternative 0 <= sum <= 1 by doing sum/2 ..
    assert 0 <= sum_wcrts and sum_wcrts <= 2  
    
    return s, sum_wcrts, is_schedulable

File: cpp/src/cost_functions.cpp
#include "cost_functions.h"
#include <iostream>

/*

    TODO rand thing ....

*/

/*
double cost_function(std::vector<Task> *taskSet) {
    std::list<Task> polling_servers; 
    std::map<std::string, int> wcrts_tt;
    std::tuple<bool, std::map<std::string, int>> is_schedulable_cost = is_schedulable_cost;
    double cost_tt = 0, cost_et = 0, cost = 0;
    bool is_schedulable;

    for (auto it : *taskSet) { 
        if (it.et_subset != NULL) { polling_servers.push_back(it); };        
    }

    for(auto it : polling_servers) {
        is_schedulable_cost = is_polling_server_schedulable(&it);
        double wcrts_et_sum = 0;
        
        if ( std::get<0>(is_schedulable_cost) ) {
            std::map<std::string, int> wcrts = std::get<1>(is_schedulable_cost);
            for (auto it : polling_servers) {
                for(int i = 0; i < it.et_subset->size(); i = i + 1) {// it.et_subset) { 
                    
                    wcrts_et_sum += double(wcrts[it.et_subset->at(i).name]); 
                    } 
                } 
        } else {
            cost_et = double(1);
            is_schedulable = false; 
        } 
    }

    wcrts_tt = edf(taskSet);
    
    if (wcrts_tt.contains("NO")) {
        cost_tt = double(1);
        //std::cout << " helllllooo" << std::endl;
        is_schedulable = false;
    } else {
        for (auto it : *taskSet) { 
            cost_tt +=  double(wcrts_tt[it.name]) / double(it.deadline); 
        }
        cost_tt *= 1/double(taskSet->size());
    }

    //std::cout << "NORMAL: " << cost_tt / taskSet->size() << std::endl;

    cost_et *= 1.0  / double(polling_servers.size());
    cost = cost_tt + cost_et;
    std::cout << "cost function cost is: " << cost << " et: " << cost_et << " tt: " << cost_tt << std::endl;
    return cost; 
}

*/

// release jobs 
void get_ready_list(std::vector<Task> *taskSet, std::list<Task> *readyList, int cycle) {  
    for(auto task : *taskSet) {
        if (cycle % task.period == 0) {
            // release new instance of task with deadline = Di + t and release time = t
            readyList->push_back(Task(task.name, task.duration, task.period, task.type, task.priority, task.deadline + cycle, cycle));
        }
    }    
}

/* deadline used here is relative deadline. section 4.6 hard real time 
   only for t1 = 0 not general case. also assumes that Di <= Ti
*/
int processor_demand(int t2, std::vector<Task> *periodic_tasks) {
    int n = periodic_tasks->size(); 
    int processor_demand=0; 
    double eta_i, period_i, deadline_i;

    // demand for some task is number of times scheduled * duration
    for(int i=0; i<n; i=i+1) {
        period_i = (double) periodic_tasks->at(i).period;
        deadline_i = (double) periodic_tasks->at(i).deadline; // line would be really long if we did this in line below
        eta_i = (t2 + period_i - deadline_i) /  period_i; // we could do without 2nd term but more general here

        processor_demand +=  ((int) eta_i) * periodic_tasks->at(i).duration;
    }

    return processor_demand;
}

//https://www.wolframalpha.com/input?i=lcm%28a%2Cb%2Cc%29+%3D%3D+lcm%28lcm%28a%2Cb%29%2C+c%29
//https://en.cppreference.com/w/cpp/numeric/lcm
//https://www.geeksforgeeks.org/lcm-of-given-array-elements/ 
std::tuple<bool, std::map<std::string, double>, std::vector<std::string>> edf(std::vector<Task> *tt_tasks) { 
    int hyperperiod = tt_tasks->front().period;
    
    // https://stackoverflow.com/questions/4210470/looping-on-c-iterators-starting-with-second-or-nth-item 
    for(auto it = std::next(tt_tasks->begin()); it != tt_tasks->end(); ++it) {
        hyperperiod = std::lcm(hyperperiod, it->period);
    }
    
    //std::cout << "hyperperiod is: " << hyperperiod << std::endl;
    std::map<std::string, double> wcrts;
    std::vector<std::string> schedule;

    if (processor_demand(hyperperiod, tt_tasks) > hyperperiod) {
        return std::tuple(false, wcrts, schedule);
    }  

    std::list<Task> ready_list; 
    int t = 0;

    // TODO: skip if none ready
    while (t < hyperperiod) {
        
        // check for deadline miss
        for (auto job : ready_list) {
            if (job.duration > 0 and job.deadline <= t) {
                return std::tuple(false, wcrts, schedule); 
            }
        }

        // release new task instances 
        get_ready_list(tt_tasks, &ready_list, t);

        // execute task instance with earliest deadline, if there are any ready tasks
        if (ready_list.size() == 0) {
            schedule.push_back("IDLE"); // no tasks instantiated, nothing to do 
        } else {

            // get task instance with earliest deadline https://stackoverflow.com/questions/26766136/how-to-get-min-or-max-element-in-a-vector-of-objects-in-c-based-on-some-field 
            auto ed_task = std::min_element(ready_list.begin(), ready_list.end(), Task::ByDeadline());  
            schedule.push_back(ed_task->name);
            ed_task->duration -= 1;

            // check if current response time larger than than current maximum
            if (ed_task->duration == 0 && ed_task->deadline >= t) {
                ready_list.erase(ed_task); // remove from ready list 
                if (!wcrts.contains(ed_task->name) || t - ed_task->release_time >= wcrts[ed_task->name]) {
                    wcrts[ed_task->name] = t - ed_task->release_time;
                }
            } else if(ed_task->duration == 0 && ed_task->deadline < t) { // this check should be redundant but to be safe. should be caught above if this is the case 
                return std::tuple(false, wcrts, schedule);  
            }
        }

        t = t + 1;
    }

    return std::tuple(true, wcrts, schedule); 
}

// determine if some polling server with a given budget, period, deadline and set of et tasks is schedulable
std::tuple<bool, std::map<std::string, double>> is_polling_server_schedulable(Task* polling_server) {
    
    int budget = polling_server->duration;
    int period = polling_server->period;
    int deadline = polling_server->deadline;
    std::vector<Task>* et_set = polling_server->et_subset;  

    int delta = period + deadline - 2*budget;
    double alpha = double(budget) / double(period);
    double supply, demand, cur_response_time; 
    std::map<std::string, double> response_times;
    

    int hyperperiod = et_set->front().period;
    
    // https://stackoverflow.com/questions/4210470/looping-on-c-iterators-starting-with-second-or-nth-item
    for(auto it = std::next(et_set->begin()); it != et_set->end(); ++it) {
        hyperperiod = std::lcm(hyperperiod, it->period);
    }


    for(auto et_it : *et_set) {
        // initialize response time of current et task to a value exceeding its deadline
        cur_response_time = et_it.deadline + 1;
        
        for(int t = 0; t <= hyperperiod; t = t + 1) {
            supply = alpha * (double(t) - double(delta));
            demand = 0;

            for(auto et_it_j : *et_set) {

                // compute maximum demand at time t
                if (et_it_j.priority >= et_it.priority) { 
                    demand = demand + std::ceil(double(t) / double(et_it_j.period)) * double( et_it_j.duration);
                }
            } // end of traversal of other ets

            if (supply >= demand) { 
                cur_response_time = t;
                response_times[et_it.name] = cur_response_time; 
                break; 
            }
        } // end of inner for

        if (cur_response_time >= et_it.deadline) { 
            return std::tuple<bool, std::map<std::string, double>>(false, response_times);
        }    

    } // end of outer for

    return std::tuple<bool, std::map<std::string, double>>(true, response_times);

}
